respy/state_space.py

```diff
@@ -612,114 +525,255 @@ def _create_dense_state_space_grid(optim_paras):
 
     for i in range(states.shape[0]):
-
-        idx_current = indexer_current[array_to_tuple(indexer_current, states[i])]
-
-        for choice in range(n_choices):
-            # Check if the state in the future is admissible.
-            if is_inadmissible[idx_current, choice]:
-                continue
-            else:
-                # Cut off the period which is not necessary for the indexer.
+        j = 0
+        for choice, is_available in enumerate(choice_set):
+            if is_available:
                 child = states[i].copy()
 
+                # Adjust period.
+                child[0] += 1
+
                 # Increment experience if it is a choice with experience
                 # accumulation.
                 if choice < n_choices_w_exp:
-                    child[choice] += 1
+                    child[choice + 1] += 1
 
                 # Change lagged choice by shifting all existing lagged choices by
                 # one period and inserting the current choice in first position.
                 if n_lagged_choices:
                     child[
-                        n_choices_w_exp + 1 : n_choices_w_exp + n_lagged_choices
-                    ] = child[n_choices_w_exp : n_choices_w_exp + n_lagged_choices - 1]
-                    child[n_choices_w_exp] = choice
+                        n_choices_w_exp + 2 : n_choices_w_exp + n_lagged_choices + 1
+                    ] = child[n_choices_w_exp + 1 : n_choices_w_exp + n_lagged_choices]
+                    child[n_choices_w_exp + 1] = choice
 
                 # Get the position of the continuation value.
-                idx_future = indexer_future[array_to_tuple(indexer_future, child)]
-                indices[idx_current, choice] = idx_future
+                idx_future = indexer[array_to_tuple(indexer, child)]
+                indices[i, j] = idx_future
+
+                j += 1
 
     return indices
 
 
-def _create_dense_state_space_covariates(dense_grid, optim_paras, options):
-    if dense_grid:
-        columns = create_dense_state_space_columns(optim_paras)
+@parallelize_across_dense_dimensions
+@nb.njit
+def _get_continuation_values(
+    core_indices,
+    dense_complex_index,
+    child_indices,
+    core_index_and_dense_vector_to_dense_index,
+    expected_value_functions,
+    n_periods,
+):
 
-        df = pd.DataFrame(data=dense_grid, columns=columns).set_index(
-            columns, drop=False
-        )
+    if len(dense_complex_index) == 3:
+        period, choice_set, dense_idx = dense_complex_index
+    elif len(dense_complex_index) == 2:
+        period, choice_set = dense_complex_index
+        dense_idx = 0
 
-        covariates = compute_covariates(df, options["covariates_dense"])
-        covariates = covariates.apply(downcast_to_smallest_dtype)
-        covariates = covariates.to_dict(orient="index")
-        covariates = convert_dictionary_keys_to_dense_indices(covariates)
+    sum_choices = 0
```

Comment:
I think I also tried np.sum before and it did not work, but now it is fine for tuples and arrays with boolean values. Can you test it again?